### PR TITLE
Allow self-referencing product blocks

### LINKED ISSUE #108: Allow self-referencing product blocks
# Problem

Product blocks can already be linked to other product blocks:
```py
# example_normal.py
from typing import Optional

from orchestrator.domain.base import ProductBlockModel
from orchestrator.types import is_of_type, is_optional_type


class Bar(ProductBlockModel, product_block_name="Bar"):
    int_field: Optional[int] = None
    str_field: Optional[str] = None


class Foo(ProductBlockModel, product_block_name="Foo"):
    sub_block: Bar
    sub_block_2: Optional[Bar] = None
    int_field: Optional[int] = None


print(Foo.__dict__["__annotations__"]["sub_block"])
# <class '__main__.Bar'>
print(Foo.__dict__["__annotations__"]["sub_block_2"])
# typing.Optional[__main__.Bar]

print(is_of_type(Foo.__dict__["__annotations__"]["sub_block"], ProductBlockModel))
# True
print(is_optional_type(Foo.__dict__["__annotations__"]["sub_block_2"], ProductBlockModel))
# True
```

But linking a product block to itself requires a forward reference to its own type, which poses a challenge with type checking.

**Option 1: from __future__ import annotations**
```py
from __future__ import annotations

from typing import Optional

from orchestrator.domain.base import ProductBlockModel
from orchestrator.types import is_of_type, is_optional_type


class Foo(ProductBlockModel, product_block_name="Foo"):
    sub_block: Foo
    sub_block_2: Optional[Foo] = None
    int_field: Optional[int] = None


print(Foo.__dict__["__annotations__"]["sub_block"])
# Foo
print(Foo.__dict__["__annotations__"]["sub_block_2"])
# Optional[Foo]

print(is_of_type(Foo.__dict__["__annotations__"]["sub_block"], ProductBlockModel))
# False
print(is_optional_type(Foo.__dict__["__annotations__"]["sub_block_2"], ProductBlockModel))
# False
```

The type is reduced to a string `"Foo"` and therefore the type check doesn't work.

**Option 2: from typing import ForwardRef**
(also possible without an import by writing the type as `sub_block: "Foo"`, but since we already `import annotations` in orchestator we need to explicitly create a forwardref)
```py
from typing import Optional, ForwardRef

from orchestrator.domain.base import ProductBlockModel
from orchestrator.types import is_of_type, is_optional_type


class Foo(ProductBlockModel, product_block_name="Foo"):
    sub_block: ForwardRef("Foo")
    sub_block_2: Optional[ForwardRef("Foo")] = None
    int_field: Optional[int] = None


print(Foo.__dict__["__annotations__"]["sub_block"])
# ForwardRef('Foo')
print(Foo.__dict__["__annotations__"]["sub_block_2"])
# typing.Optional[ForwardRef('Foo')]

print(is_of_type(Foo.__dict__["__annotations__"]["sub_block"], ProductBlockModel))
# False
print(is_optional_type(Foo.__dict__["__annotations__"]["sub_block_2"], ProductBlockModel))
# False
```

The type is now a `ForwardRef('Foo')`, which isn't much better as it still fails the type check.

# Proposal

Add `is_forward_ref()` to `orchestrator.types` which does something along the lines of:

```py
import re

RE_SAFE_TYPE = re.compile("[a-z0-9_]+", flags=re.IGNORECASE)

def is_forward_ref(t, test_type):
    if not isinstance(t, ForwardRef):
        return False
    try:
        type_name = t.__forward_arg__
    except AttributeError:
        return False

    if isinstance(type_name, str) and RE_SAFE_TYPE.fullmatch(type_name):
        return issubclass(eval(type_name), test_type)
    return False


print(is_forward_ref(Foo.__dict__["__annotations__"]["sub_block"], ProductBlockModel))
# True
```

### PR DESCRIPTION
**Pull request by Mark90**
Closes #108 by using `get_type_hints()` which evaluates any `ForwardRef` types to the actual type

### CODE REVIEW

File: test/unit_tests/domain/test_base.py [hunk 1]
```diff
@@ -44,6 +48,131 @@ def test_product_block_metadata(test_product_block_one, test_product_one, test_p
...
+    # Load block 11 and verify the nested blocks still exist
+    newmodel11 = ProductTypeOneNestedForTest.from_subscription(model11.subscription_id)
+    assert newmodel11.block.int_field == 11
+    assert newmodel11.block.sub_block.int_field == 20
+    assert newmodel11.block.sub_block.sub_block.int_field == 30
+    assert newmodel11.block.sub_block.sub_block.sub_block is None
+
+
+# TODO add test for List of nested blocks
+# def test_product_block_nested_list(test_product_model_nested, test_product_type_one_nested):
```
Comment (Mark90) [inline test/unit_tests/domain/test_base.py:173]: Add this testcase

**Comment by pboers1988** [comment]
Looks good 👍🏼

File: orchestrator/services/processes.py [hunk 2]
```diff
@@ -64,6 +64,14 @@ def get_thread_pool() -> ThreadPoolExecutor:
     return _workflow_executor
 
 
+def shutdown_thread_pool() -> None:
```
Comment (pboers1988) [inline orchestrator/services/processes.py:67]: Is this only called in the tests?

**Review by pboers1988 — APPROVED** [review]

File: test/unit_tests/domain/conftest.py [hunk 3]
```diff
@@ -1,7 +1,15 @@
+from test.unit_tests.domain.products.product_blocks.product_block_list_nested import (  # noqa: F401  # noqa: F401
```
Comment (tjeerddie) [inline test/unit_tests/domain/conftest.py:1]: double noqa's

**Review by tjeerddie — APPROVED** [review]

**Pull request merged**